bin/faucet/src/faucet/mod.rs

```diff
@@ -0,0 +1,422 @@
+
+use anyhow::{Context, anyhow};
+use miden_lib::{note::create_p2id_note, transaction::TransactionKernel};
+use miden_objects::{
+    AccountError, Digest, Felt, NoteError, TransactionScriptError,
+    account::{Account, AccountDelta, AccountFile, AccountId, AuthSecretKey},
+    asset::FungibleAsset,
+    block::BlockNumber,
+    crypto::{
+        merkle::{MmrPeaks, PartialMmr},
+        rand::{FeltRng, RpoRandomCoin},
+    },
+    note::Note,
+    transaction::{
+        ChainMmr, ExecutedTransaction, ProvenTransaction, TransactionArgs, TransactionScript,
+    },
+    vm::AdviceMap,
+};
+use miden_tx::{
+    LocalTransactionProver, ProvingOptions, TransactionExecutor, TransactionExecutorError,
+    TransactionProver, TransactionProverError, auth::BasicAuthenticator,
+    utils::parse_hex_string_as_word,
+};
+use rand::{random, rngs::StdRng};
+use serde::Serialize;
+use store::FaucetDataStore;
+use tokio::sync::{mpsc, oneshot};
+use tonic::Code;
+use tracing::{info, instrument};
+
+use crate::{
+    rpc_client::{RpcClient, RpcError},
+    types::{AssetAmount, NoteType},
+};
+
+mod store;
+
+pub const DISTRIBUTE_FUNGIBLE_ASSET_SCRIPT: &str = include_str!("distribute_fungible_asset.masm");
+
+// FAUCET CLIENT
+// ================================================================================================
+
+/// The faucet's account ID.
+///
+/// Used as a type safety mechanism to avoid confusion with user account IDs,
+/// and allows us to implement traits.
+#[derive(Clone, Copy)]
+pub struct FaucetId(AccountId);
+
+impl FaucetId {
+    pub fn inner(self) -> AccountId {
+        self.0
+    }
+}
+
+impl std::fmt::Display for FaucetId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl Serialize for FaucetId {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_str(&self.0.to_hex())
+    }
+}
+
+/// A request for minting to the [`Faucet`].
+pub struct MintRequest {
+    /// Destination account.
+    pub account_id: AccountId,
+    /// Whether to generate a public or private note to hold the minted asset.
+    pub note_type: NoteType,
+    /// The amount to mint.
+    pub asset_amount: AssetAmount,
+}
+
+type MintResult<T> = Result<T, MintError>;
+
+/// Error indicating what went wrong in the minting process for a request.
+#[derive(Debug, thiserror::Error)]
+pub enum MintError {
+    #[error("building the p2id note failed")]
+    BuildingP2IdNote(#[source] NoteError),
+    #[error("compiling the tx script failed")]
+    ScriptCompilation(#[source] TransactionScriptError),
+    #[error("execution of the tx script failed")]
+    Execution(#[source] TransactionExecutorError),
+    #[error("proving the tx failed")]
+    Proving(#[source] TransactionProverError),
+    #[error("submitting the tx to the node failed")]
+    Submission(#[source] RpcError),
+}
```

Comment:
I'm unsure how many of these can actually error?

For example, I feel like the note build and compilation should always succeed. But I was unsure so I err'd on the side of caution.